Core/Parser/SchemaParser.cs

```diff
@@ -207,7 +237,42 @@ public async Task<ISchema> Parse()
                 {
                     throw new UnrecognizedTypeException(typeToken, definitionToken.Lexeme);
                 }
+                var reference = _definitions[typeToken.Lexeme];
+                var referenceScope = reference.Scope;
+                var definition = _definitions[definitionToken.Lexeme];
+                var definitionScope = definition.Scope;
+                // You're not allowed to reference types declared within a union from elsewhere
+                // Check if reference has a union in scope but definition does not have the same union in scope
+                // Throw ReferenceScopeException if so
+
+                // It might be better for this to go inside BebopSchema.Validate but it's much simpler if I can use the typeReferences map
+                if (referenceScope is not null && referenceScope.Find((parent) => parent is UnionDefinition) is UnionDefinition union)
+                {
+                    if (definitionScope is null || !definitionScope.Contains(union))
+                    {
+                        throw new ReferenceScopeException(definition, reference, "union");
+                    }
+                }
             }
+            /*
+            foreach (var definition in _definitions.Values)
+            {
+                var scope = definition.Scope;
+                var dependencies = definition.Dependencies();
+
+                
+                foreach (var dependency in dependencies)
+                {
+                    var depScope = _definitions[dependency].Scope;
+                    if (depScope is not null && depScope.Find((parent) => parent is UnionDefinition) is UnionDefinition union) {
+                        if (scope is null || !scope.Contains(union))
+                        {
+                            throw new ReferenceScopeException(definition, _definitions[dependency], "union");
+                        }
+                    }
+                }
+                
+            }*/
```

Comment:
My classic review comment about commented-out code:

If there's a good reason to keep it around but commented-out, add a comment explaining why.

If there isn't a good reason, let's get rid of it.